<a href="https://colab.research.google.com/github/hari-prasath-j/Project.2_Spam_or_ham/blob/main/Project_2_Spam_or_ham.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# SMS: Spam or Ham:

For this project, I chose the SMS Spam Collection Dataset.

It contains the text of 5572 SMS messages and a label, classifying the message as "spam" or "ham" using naive bayes classifier.


In [ ]:
# Find the encoding format
with open('/content/spam.csv') as spam:
  print(spam,type(spam))

<_io.TextIOWrapper name='/content/spam.csv' mode='r' encoding='UTF-8'> <class '_io.TextIOWrapper'>


In [ ]:
#load the dataset
import pandas as pd
df= pd.read_csv("/content/spam.csv",encoding='latin-1')

In [ ]:
df

v1                                                 v2 Unnamed: 2  \
0      ham  Go until jurong point, crazy.. Available only ...        NaN   
1      ham                      Ok lar... Joking wif u oni...        NaN   
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3      ham  U dun say so early hor... U c already then say...        NaN   
4      ham  Nah I don't think he goes to usf, he lives aro...        NaN   
...    ...                                                ...        ...   
5567  spam  This is the 2nd time we have tried 2 contact u...        NaN   
5568   ham              Will Ì_ b going to esplanade fr home?        NaN   
5569   ham  Pity, * was in mood for that. So...any other s...        NaN   
5570   ham  The guy did some bitching but I acted like i'd...        NaN   
5571   ham                         Rofl. Its true to its name        NaN   

     Unnamed: 3 Unnamed: 4  
0           NaN        NaN  
1           NaN        NaN  
2           NaN        NaN  
3           NaN        NaN  
4           NaN        NaN  
...         ...        ...  
5567        NaN        NaN  
5568        NaN        NaN  
5569        NaN        NaN  
5570        NaN        NaN  
5571        NaN        NaN  

[5572 rows x 5 columns]

In [ ]:
#renameing the columns
df.rename(columns={"v1":"label","v2":"messege"},inplace=True)
df

label                                            messege Unnamed: 2  \
0      ham  Go until jurong point, crazy.. Available only ...        NaN   
1      ham                      Ok lar... Joking wif u oni...        NaN   
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3      ham  U dun say so early hor... U c already then say...        NaN   
4      ham  Nah I don't think he goes to usf, he lives aro...        NaN   
...    ...                                                ...        ...   
5567  spam  This is the 2nd time we have tried 2 contact u...        NaN   
5568   ham              Will Ì_ b going to esplanade fr home?        NaN   
5569   ham  Pity, * was in mood for that. So...any other s...        NaN   
5570   ham  The guy did some bitching but I acted like i'd...        NaN   
5571   ham                         Rofl. Its true to its name        NaN   

     Unnamed: 3 Unnamed: 4  
0           NaN        NaN  
1           NaN        NaN  
2           NaN        NaN  
3           NaN        NaN  
4           NaN        NaN  
...         ...        ...  
5567        NaN        NaN  
5568        NaN        NaN  
5569        NaN        NaN  
5570        NaN        NaN  
5571        NaN        NaN  

[5572 rows x 5 columns]

In [ ]:
#drop the uncessery columns
df.drop(['Unnamed: 2', 'Unnamed: 3', 'Unnamed: 4'],inplace=True,axis = 1)
df

label                                            messege
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...
...    ...                                                ...
5567  spam  This is the 2nd time we have tried 2 contact u...
5568   ham              Will Ì_ b going to esplanade fr home?
5569   ham  Pity, * was in mood for that. So...any other s...
5570   ham  The guy did some bitching but I acted like i'd...
5571   ham                         Rofl. Its true to its name

[5572 rows x 2 columns]

In [ ]:
#Class distribution prior
df.label.value_counts()

ham     4825
spam     747
Name: label, dtype: int64

In [ ]:
#covert string to num
df["label"]=df["label"].apply(lambda x: 1 if x == "spam" else 0)

In [ ]:
df


label                                            messege
0         0  Go until jurong point, crazy.. Available only ...
1         0                      Ok lar... Joking wif u oni...
2         1  Free entry in 2 a wkly comp to win FA Cup fina...
3         0  U dun say so early hor... U c already then say...
4         0  Nah I don't think he goes to usf, he lives aro...
...     ...                                                ...
5567      1  This is the 2nd time we have tried 2 contact u...
5568      0              Will Ì_ b going to esplanade fr home?
5569      0  Pity, * was in mood for that. So...any other s...
5570      0  The guy did some bitching but I acted like i'd...
5571      0                         Rofl. Its true to its name

[5572 rows x 2 columns]

In [ ]:
#training the model
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(df["messege"],df["label"],test_size=0.3,random_state=42)

In [ ]:
y_train.value_counts()

0    3372
1     528
Name: label, dtype: int64

In [ ]:
y_test.value_counts()

0    1453
1     219
Name: label, dtype: int64

# Count Vectorizer(bag of words)
transform the messege columns string to numbers

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
vector= CountVectorizer(lowercase=True, #Convert to lowercase
                        stop_words='english'#remove the stop words
                        )

In [ ]:
x_train_transformed = vector.fit_transform(x_train) # gives vector for x_train
x_test_transformed = vector.transform(x_test) # gives vector for x_test

In [ ]:
x_train_transformed

<3900x6946 sparse matrix of type '<class 'numpy.int64'>'
	with 30413 stored elements in Compressed Sparse Row format>

In [ ]:
x_train_transformed.toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [ ]:
vector.get_feature_names_out()

array(['00', '000', '000pes', ..., 'ûïharry', 'ûò', 'ûówell'],
      dtype=object)

# Model Training

In [ ]:
from sklearn.naive_bayes import MultinomialNB
classifier = MultinomialNB()
classifier.fit(x_train_transformed, y_train)

MultinomialNB()

Prediction on datasets

In [ ]:
y_test_Predicted = classifier.predict(x_test_transformed)

In [ ]:
y_test_Predicted

array([0, 0, 1, ..., 0, 0, 0])

In [ ]:
y_test.count()

1672

# Model Evaluation

In [ ]:
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report

In [ ]:
print("Accuracy_score:",accuracy_score(y_test, y_test_Predicted))

Accuracy_score: 0.9832535885167464


In summary, an accuracy score of 0.9832535885167464 suggests that the model performed very well in making predictions on the dataset it was tested on, accurately classifying approximately 98.32% of the instances

In [ ]:
results= confusion_matrix(y_test, y_test_Predicted)
print(results)

[[1442   11]
 [  17  202]]


The confusion matrix helps in understanding where the model is making mistakes. It's particularly useful for assessing whether the model is more prone to certain types of errors, like false positives or false negatives. These insights can guide further model optimization or fine-tuning, depending on the specific objectives of the classification task.

In our case our model was

* TP: 1442 (Predicted Positive, Actually Positive)
* TN: 202 (Predicted Negative, Actually Negative)
* FP: 11 (Predicted Positive, Actually Negative)
* FN: 17 (Predicted Negative, Actually Positive)



In [ ]:
print (classification_report(y_test, y_test_Predicted))

              precision    recall  f1-score   support

           0       0.99      0.99      0.99      1453
           1       0.95      0.92      0.94       219

    accuracy                           0.98      1672
   macro avg       0.97      0.96      0.96      1672
weighted avg       0.98      0.98      0.98      1672



This classification report provides a detailed view of how well the model performs in terms of precision, recall, and overall accuracy for each class and across the entire dataset.

* Precision: Precision is a measure of how many instances predicted as positive are actually positive.

   *    For class 0, the precision is 0.99, indicating that 99% of the instances predicted as class 0 were indeed class 0.
   * For class 1, the precision is 0.95, meaning that 95% of the instances predicted as class 1 were actually class 1.
* Recall (Sensitivity): Recall is a measure of how many actual positive instances were correctly predicted by the model.

 * For class 0, the recall is 0.99, indicating that 99% of the actual class 0 instances were correctly predicted by the model.
 * For class 1, the recall is 0.92, meaning that 92% of the actual class 1 instances were correctly predicted by the model.
* F1-score: The F1-score is the harmonic mean of precision and recall, providing a balance between the two metrics.

 * For class 0, the F1-score is 0.99, reflecting a high balance between precision and recall.
 * For class 1, the F1-score is 0.94, which also represents a reasonably balanced performance in precision and recall.
* Support: This indicates the number of instances of each class in the dataset.

 * There are 1453 instances of class 0 and 219 instances of class 1.

* Accuracy: The overall accuracy of the model across both classes is 0.98, meaning that the model correctly predicted 98% of all instances.

* Macro Avg: This is the average precision, recall, and F1-score calculated across classes. The macro average precision, recall, and F1-score are 0.97, 0.96, and 0.96, respectively.

* Weighted Avg: This is the weighted average of precision, recall, and F1-score, considering the support (number of instances) for each class. The weighted average precision, recall, and F1-score are 0.98, 0.98, and 0.98, respectively.

# Testing Unknown messege

In [ ]:
New_mail= "Please call me you WON a prize of 10000 dollars!"

In [ ]:
New_mail_transformed=(vector.transform([New_mail]))

In [ ]:
classifier.predict(New_mail_transformed)

array([1])

Our model correctly predicted...